File: leetcode/tree/98. Validate Binary Search Tree/js/main.js
// Inorder Traversal
// Time O(N)
// Space O(N)
const isValidBST = function (root) {
  let stack = [];
  let node = root;
  let inorder = -Number.MAX_VALUE;

  while (stack.length || node !== null) {
    while (node !== null) {
      stack.push(node);
      node = node.left;
    }

    node = stack.pop();

    if (inorder >= node.val) {
      return false;
    }

    inorder = node.val;
    node = node.right;
  }

  return true;
};

// Time O(N)
// Space O(N)
const isValidBST_II = root => {
  let node = dfs(root);
  return node.isBST;

  function dfs(node) {
    if (node == null) return new MinMax();

    let left = dfs(node.left);
    let right = dfs(node.right);

    let m = new MinMax();

    if (!left.isBST || !right.isBST || left.max >= node.val || right.min <= node.val) {
      m.isBST = false;
      return m;
    }

    // for min take min value from left
    // for max take max value from right
    m.min = node.left != null ? left.min : node.val;
    m.max = node.right != null ? right.max : node.val;

    return m;
  }
};

class MinMax {
  constructor(min, max, isBST) {
    this.min = min || Number.MAX_VALUE;
    this.max = max || -Number.MAX_VALUE;
    this.isBST = isBST || true;
  }
}

File: leetcode/tree/98. Validate Binary Search Tree/py/main.py
class Solution:
    # Time O(N)
    # Space O(N)
    def isValidBST(self, root):
        if not root:
            return False

        def isLeaf(node):
            return not node.left and not node.right

        def dfs(node):
            left, right = None, None

            if isLeaf(node):
                return Node(node.val, node.val, True)

            if node.left:
                left = dfs(node.left)

            if node.right:
                right = dfs(node.right)

            if (right and not right.is_valid) or (left and not left.is_valid):
                return Node()

            if left and right:
                if node.val > left.max and node.val < right.min:
                    return Node(left.min, right.max, True)
                return Node()

            if left and node.val > left.max:
                return Node(left.min, node.val, True)

            if right and node.val < right.min:
                return Node(node.val, right.max, True)

            return Node()

        return dfs(root).is_valid


class Node:
    def __init__(self, min=-1, max=-1, is_valid=False):
        self.min = min
        self.max = max
        self.is_valid = is_valid